### PR TITLE
Allow unlimited maximum pool size

### LINKED ISSUE #600: [BUG]Maximum pool size exceeded error with managed memory
**Describe the bug**

By default, if I don't set `maximum_pool_size ` we seem to enforce a hard limit even with managed memory and throw a `Maximum pool size exceeded` error. If I set to a large value (>GPU memory) , things seem to work as expected. 

**Steps/Code to reproduce bug**
Example
```python
import rmm
import cudf
rmm.reinitialize(pool_allocator=True,initial_pool_size=15e+9,managed_memory=False)
s = cudf.Series([1]).repeat(1_000_000_000)
```
```python
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
<ipython-input-2-05bad8b65c06> in <module>
      2 import cudf
      3 rmm.reinitialize(pool_allocator=True,initial_pool_size=15e+9,managed_memory=False)
----> 4 s = cudf.Series([1]).repeat(1_000_000_000)

/nvme/0/vjawa/conda/envs/tpcx-bb-oct-7/lib/python3.7/site-packages/cudf/core/frame.py in repeat(self, repeats, axis)
   1583             )
   1584 
-> 1585         return self._repeat(repeats)
   1586 
   1587     def _repeat(self, count):

/nvme/0/vjawa/conda/envs/tpcx-bb-oct-7/lib/python3.7/site-packages/cudf/core/frame.py in _repeat(self, count)
   1590 
   1591         result = self.__class__._from_table(
-> 1592             libcudf.filling.repeat(self, count)
   1593         )
   1594 

cudf/_lib/filling.pyx in cudf._lib.filling.repeat()

cudf/_lib/filling.pyx in cudf._lib.filling._repeat_via_size_type()

MemoryError: std::bad_alloc: RMM failure at:../include/rmm/mr/device/pool_memory_resource.hpp:189: Maximum pool size exceeded


```
**Expected behavior**
If I set it to a high value we get the same expected behavior . 

```python
import rmm
import cudf
rmm.reinitialize(pool_allocator=True,initial_pool_size=15e+9,managed_memory=True,maximum_pool_size=30e+9)
s = cudf.Series([1]).repeat(1_000_000_000)
print(len(s))
```
```python
1000000000
```
**Environment details (please complete the following information):**
 ```
librmm                    0.16.0a201007   cuda10.2_g273f92d_397    rapidsai-nightly
rmm                       0.16.0a201007   cuda_10.2_py37_g273f92d_397    rapidsai-nightly
cudf                      0.16.0a201007   cuda_10.2_py37_g8950c1a69e_1962    rapidsai-nightly
dask-cudf                 0.16.0a201007   py37_g8950c1a69e_1962    rapidsai-nightly
libcudf                   0.16.0a201007   cuda10.2_g8950c1a69e_1962    rapidsai-nightly
```

CC: @harrism , @ayushdg

### PR DESCRIPTION
**Pull request by harrism**
Fixes #600 

In some cases you want a pool with no maximum size -- e.g. when your upstream resource is managed memory. This PR allows a truly optional maximum pool size and improves the pool growth heuristic to grow exponentially in this case, with exponential backoff in case of failure. It will retry smaller allocations until either it succeeds or allocating the requested size from upstream fails.

Also changes to using `thrust::optional` for the initial and maximum pool size parameters to `pool_memory_resource`.  (std::optional can be used in the future when C++17 is our minimum)

### CODE REVIEW

**Comment by VibhuJawa** [comment]
So i tried a similar example on the branch and am still getting similar failures. See below: 

### Failure if don't set `max_memory`
```python
In [1]: import cudf
   ...: import numpy as np
   ...: import rmm

In [2]: rmm.reinitialize(pool_allocator=True,initial_pool_size=15e+9,managed_memory=False)

In [3]: s = cudf.Series(np.ones(1_000_000_000))
Trying 15000000000 B
Trying 8000000000 B
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
<ipython-input-3-ea7dbdbb7bc3> in <module>
----> 1 s = cudf.Series(np.ones(1_000_000_000))

/nvme/0/vjawa/conda/envs/rapids-tpcx-bb-oct-9-2.yml/lib/python3.7/site-packages/cudf/core/series.py in __init__(self, data, index, dtype, name, nan_as_null)
    186 
    187         if not isinstance(data, column.ColumnBase):
--> 188             data = column.as_column(data, nan_as_null=nan_as_null, dtype=dtype)
    189         else:
    190             if dtype is not None:

/nvme/0/vjawa/conda/envs/rapids-tpcx-bb-oct-9-2.yml/lib/python3.7/site-packages/cudf/core/column/column.py in as_column(arbitrary, nan_as_null, dtype, length)
   1725             data = as_column(
   1726                 cupy.asarray(arbitrary, dtype=arb_dtype),
-> 1727                 nan_as_null=nan_as_null,
   1728             )
   1729         else:

/nvme/0/vjawa/conda/envs/rapids-tpcx-bb-oct-9-2.yml/lib/python3.7/site-packages/cudf/core/column/column.py in as_column(arbitrary, nan_as_null, dtype, length)
   1565         elif np.issubdtype(col.dtype, np.floating):
   1566             if nan_as_null or (mask is None and nan_as_null is None):
-> 1567                 mask = libcudf.transform.nans_to_nulls(col.fillna(np.nan))
   1568                 col = col.set_mask(mask)
   1569         elif np.issubdtype(col.dtype, np.datetime64):

/nvme/0/vjawa/conda/envs/rapids-tpcx-bb-oct-9-2.yml/lib/python3.7/site-packages/cudf/core/column/numerical.py in fillna(self, fill_value)
    441             else:
    442                 fill_value = fill_value.astype(self.dtype)
--> 443         result = libcudf.replace.replace_nulls(self, fill_value)
    444 
    445         return result

cudf/_lib/replace.pyx in cudf._lib.replace.replace_nulls()

cudf/_lib/replace.pyx in cudf._lib.replace.replace_nulls_scalar()

MemoryError: std::bad_alloc: RMM failure at:../include/rmm/mr/device/pool_memory_resource.hpp:194: Maximum pool size exceeded
```


### No Failure if set `max_memory`

```python
In [1]: import cudf
   ...: import numpy as np
   ...: import rmm

In [2]: rmm.reinitialize(pool_allocator=True,initial_pool_size=15e+9,managed_memory=True,maximum_pool_size=30e+9)

In [3]: s = cudf.Series(np.ones(1_000_000_000))
Trying 8000000000 B

In [4]: len(s)
Out[4]: 1000000000
```

**Comment by harrism** [comment]
Your second example is with managed=true. How much memory on your GPU?

Looks like I forgot to remove some print statements.

**Comment by VibhuJawa** [comment]
> Your second example is with managed=true, first is not. How much memory on your GPU?
> 
> The series you are creating needs 16GB. You init to 15GB so on a 16GB GPU there is no way to succeed.
> 
> Looks like I forgot to remove some print statements.

You are right it works with managed. Really sorry, don't know how i missed that :disappointed_relieved: . I really need to catch up on some sleep. Sorry again.  

```python
In [1]: import cudf
   ...: import numpy as np
   ...: import rmm

In [2]: rmm.reinitialize(pool_allocator=True,initial_pool_size=15e+9,managed_memory=True)

In [3]: s = cudf.Series(np.ones(1_000_000_000))
Trying 15000000000 B

In [4]: len(s)
Out[4]: 1000000000
```

File: include/rmm/mr/device/pool_memory_resource.hpp [hunk 1]
```diff
@@ -181,18 +174,49 @@ class pool_memory_resource final
...
-    if (grow_size == 0) {
-      RMM_LOG_ERROR("[A][Stream {}][Upstream {}B][FAILURE max pool size exceeded]",
-                    reinterpret_cast<void*>(stream),
-                    size);
-      RMM_FAIL("Maximum pool size exceeded", rmm::bad_alloc);
+    // if maximum size is not set
+    auto try_size =
+      maximum_pool_size_.has_value() ? size_to_grow(size) : std::max(size, pool_size());
+    while (try_size >= size) {
+      std::cout << "Trying " << try_size << " B" << std::endl;
```
Comment (harrism) [inline include/rmm/mr/device/pool_memory_resource.hpp:181]:
```suggestion
```

File: include/rmm/mr/device/pool_memory_resource.hpp [hunk 2]
```diff
@@ -181,18 +174,48 @@ class pool_memory_resource final
...
+        try_size = std::max(size, try_size / 2);
+      }
     }
-
-    current_pool_size_ += grow_size;
-    return block_from_upstream(grow_size, stream);
+    RMM_LOG_ERROR("[A][Stream {}][Upstream {}B][FAILURE max pool size exceeded]",
+                  reinterpret_cast<void*>(stream),
+                  size);
+    RMM_FAIL("Maximum pool size exceeded", rmm::bad_alloc);
```
Comment (cwharris) [inline include/rmm/mr/device/pool_memory_resource.hpp:193]: This code is a candidate for unit testing in isolation. If we template `expand_pool` to accept a `block_from_upstream` functor and extract it from the class, we'd be able to inject specific sizes in to the unit test. The class would call said extracted function.

```c++
template<typename GetBlockFunctor>
block_type expand_bool(GetBlockFunctor get_block, size_t size, freelist& blocks, cudaStream_t stream) {
  // ... real implementation as a pure function. can be tested in isolation ...
}

struct ... {
  block_type expand_bool(size_t size, freelist& blocks, cudaStream_t stream) {
    return expand_bool_static([this](...){ return this->block_from_upstream(...); }, size, blocks, stream);
  }
};
```

File: include/rmm/detail/aligned.hpp [hunk 3]
```diff
@@ -71,6 +71,20 @@ constexpr std::size_t align_down(std::size_t v, std::size_t align_bytes) noexcep
...
 
+/**
+ * @brief Checks whether a value is aligned to a multiple of a specified power of 2
+ *
+ * @param[in] v value to check for alignment
+ * @param[in] alignment amount, in bytes, must be a power of 2
+ *
+ * @return true if aligned
+ */
+constexpr bool is_aligned(std::size_t v, std::size_t align_bytes) noexcept
```
Comment (cwharris) [inline include/rmm/detail/aligned.hpp:82]: `mr_test.hpp` has an `is_aligned`. Should they be consolidated? (currently the implementations differ)

File: include/rmm/mr/device/pool_memory_resource.hpp [hunk 4]
```diff
@@ -87,34 +87,27 @@ class pool_memory_resource final
...
     std::tie(free, total) = detail::available_device_memory();
 
-    initial_pool_size = (initial_pool_size == default_initial_size)
-                          ? rmm::detail::align_up(std::min(free, total / 2), allocation_alignment)
-                          : initial_pool_size;
-    current_pool_size_ = initial_pool_size;
-    maximum_pool_size_ = (maximum_pool_size == default_maximum_size)
-                           ? rmm::detail::align_down(free, allocation_alignment)
-                           : maximum_pool_size;
+    current_pool_size_ = initial_pool_size.value_or(
```
Comment (cwharris) [inline include/rmm/mr/device/pool_memory_resource.hpp:105]: Is `initial_pool_size` guaranteed to be aligned? (does it matter?)

**Review by cwharris — CHANGES_REQUESTED** [review]

File: include/rmm/mr/device/pool_memory_resource.hpp [hunk 5]
```diff
@@ -181,18 +174,48 @@ class pool_memory_resource final
...
+  /**
+   * @brief Given a minimum size, computes an appropriate size to grow the pool.
+   *
+   * Strategy is to try to grow the pool by half the difference between
+   * the configured maximum pool size and the current pool size.
+   *
+   * @param size The size of the minimum allocation immediately needed
+   * @return size_t The computed size to grow the pool.
+   */
+  size_t size_to_grow(size_t size) const
```
[the remark this answers is not in the document]
Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp:205]: I would prefer to just get rid of it. As for static, it depends on instance state, so no.

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 4 above)
Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp:105]: Keep reading...

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 2 above)
Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp:193]: I thought the wisdom was not to unit test implementation details, but rather interfaces.

**Comment by harrism** [comment]
Moving to 0.17 branch.

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 2 above)
Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp:193]: In any case, this function depends on the state of the class, so I don't think it can be extracted cleanly.

File: include/rmm/detail/aligned.hpp (hunk 3 above)
Comment (harrism) [reply include/rmm/detail/aligned.hpp:82]: Good catch. The test one tests pointers, so I renamed it to `is_pointer_aligned()`.

File: include/rmm/mr/device/pool_memory_resource.hpp [hunk 6]
```diff
@@ -181,18 +174,53 @@ class pool_memory_resource final
...
+                  reinterpret_cast<void*>(stream),
+                  size);
+    RMM_FAIL("Maximum pool size exceeded", rmm::bad_alloc);
   }
 
+  /**
+   * @brief Given a minimum size, computes an appropriate size to grow the pool.
+   *
+   * Strategy is to try to grow the pool by half the difference between
+   * the configured maximum pool size and the current pool size.
```
Comment (jrhemstad) [inline include/rmm/mr/device/pool_memory_resource.hpp:202]:
```suggestion
   * the configured maximum pool size and the current pool size.
   *
   * Returns 0 if the requested size cannot be satisfied. 
   *
```

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 2 above)
Comment (cwharris) [reply include/rmm/mr/device/pool_memory_resource.hpp]: > I thought the wisdom was not to unit test implementation details, but rather interfaces.

I thought you hadn't read any Uncle Bob. In that case, what interface are you testing to ensure this new combination of loop, arithmetic, state, bit-alignment, exception handling, and stream processing is correct?

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (cwharris) [reply include/rmm/mr/device/pool_memory_resource.hpp:207]: by "get rid of it" do you mean move the logic in to `expand_pool`?

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 2 above)
Comment (cwharris) [reply include/rmm/mr/device/pool_memory_resource.hpp]: > I thought the wisdom was not to unit test implementation details, but rather interfaces.

This is true. But it appears we have a different understanding of this statement.

Tests should be written such that they contain no knowledge of the underlying implementation - only the expected inputs and outputs, and suspected edge cases (what should it do if I give it +/- inf, 0, or null?). 

This is testing an implementation detail:
```c++
int add_unless_special(int a, int b) { return a == 10 || b == 10 ? 0 : a + b; }

assert_eq(5, add_unless_special(3, 2));
assert_eq(0, add_unless_special(10, 10));
```
This is testing the same logic without regards to implementation details:
```c++
template<typename UnaryPredicate>
int add_unless_special_impl(UnaryPredicate is_special, int a, int b) {
  return is_special(a) || is_special(b) ? 0 : a + b;
  }
}

struct predicate
{
  int special_number;
  operator()(int x) { return x == special_number; }
}

struct constant_predicate
{
  bool result;
  operator()(int x) { return result; }
}

assert_eq(0, add_unless_special(predicate{3}, 3, 2));
assert_eq(0, add_unless_special(predicate{4}, 1, 4));
assert_eq(4, add_unless_special(predicate{7}, 8, -4));
assert_eq(0, add_unless_special(constant_predicate{false}, 1, 2));
assert_eq(3, add_unless_special(constant_predicate{true}, 1, 2));
assert_eq(0, add_unless_special(constant_predicate{false}, 3, 4));
assert_eq(7, add_unless_special(constant_predicate{true}, 3, 4));
```

The implementation detail is "a number is special if it is 10". The interface is "add these unless they are special". By injecting the meaning of "special" from within the test itself, we are no longer testing the meaning of "special", and instead only testing what happens if inputs are or are not special.

This is a contrived example, but when we take the same approach with something like `expand_pool`, we are able to test `expand_pool` as it's _own interface_, which is then used as an _implementation detail_ of `pool_memory_resource` (which can be tested without regard to `expand_pool`).
```
```

Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp]: I don't understand why a *private* function of a class that depends on instance data of that class is "too complex" if it can't be made static.  I think this request is distracting from the point of the PR.

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 6 above)
Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp:202]: Done.

**Review by jrhemstad — APPROVED** [review]

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 2 above)
Comment (jrhemstad) [reply include/rmm/mr/device/pool_memory_resource.hpp]: > > I thought the wisdom was not to unit test implementation details, but rather interfaces.
> 
> This is true. But it appears we have a different understanding of this statement.
> 
> Tests should be written such that they contain no knowledge of the underlying implementation - only the expected inputs and outputs, and suspected edge cases (what should it do if I give it +/- inf, 0, or null?).

The usual understanding of "black box" testing is that public interfaces are tested and ensure they behave in the way they are documented. Anything below that is usually considered an implementation detail. It would be impractical to test every internal function.

We already have tests that force pool growth and ensure the resource behaves as expected. That should be sufficient.

File: include/rmm/mr/device/pool_memory_resource.hpp [hunk 7]
```diff
@@ -181,18 +174,56 @@ class pool_memory_resource final
...
+      if (aligned_size <= remaining / 2) {
+        return remaining / 2;
+      } else if (aligned_size <= remaining) {
+        return aligned_size;
+      } else {
+        return 0;
+      }
+    } else
+      return std::max(size, pool_size());
+  };
```
Comment (codereport) [inline include/rmm/mr/device/pool_memory_resource.hpp:225]: If i didn't mess up my logic, this can be simplified to the following:
```suggestion
  size_t size_to_grow(size_t size) const
  {
    if (maximum_pool_size_.has_value()) {
      auto const pool_size_delta = maximum_pool_size_.value() - pool_size();
      auto const remaining       = rmm::detail::align_up(pool_size_delta, allocation_alignment);
      auto const aligned_size    = rmm::detail::align_up(size, allocation_alignment);
      return aligned_size <= remaining ? std::max(remaining / 2, aligned_size) : 0;
    } else
      return std::max(size, pool_size());
  };
```

File: include/rmm/mr/device/pool_memory_resource.hpp [hunk 8]
```diff
@@ -181,18 +174,56 @@ class pool_memory_resource final
...
-                    size);
-      RMM_FAIL("Maximum pool size exceeded", rmm::bad_alloc);
+    // Strategy. If maximum_pool_size_ is set, then grow geometrically, e.g. by halfway to the limit
+    // each time. If it is not set, grow exponentially, e.g. by doubling the pool size each time.
+    // Upon failure, attempt to back off exponentially, e.g. by half the attempted size, until
+    // either success or the attempt is less than the requested size.
+    auto try_size = size_to_grow(size);
+    while (try_size >= size) {
+      try {
+        auto b = block_from_upstream(try_size, stream);
```
Comment (codereport) [inline include/rmm/mr/device/pool_memory_resource.hpp:184]: Might add a comment like this to make it clear why there is a `+= try_size` happening after `b = ` and not before.
```suggestion
        auto b = block_from_upstream(try_size, stream);  // has side-effect: can throw
```

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 7 above)
Comment (cwharris) [reply include/rmm/mr/device/pool_memory_resource.hpp:225]: > The fact that it is wrapping functions with side-effects (aka throws) makes it pretty tricky.

fwiw, Mark mentioned (and I'm extrapolating) that it's possible (though not pretty) to get around this by wrapping said function call in another function that performs the try/catch and translates the error to another form, such as a `pair<error, result>`.

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 8 above)
Comment (jrhemstad) [reply include/rmm/mr/device/pool_memory_resource.hpp:184]: Since we're already using `thrust::optional`, maybe `block_from_upstream` can be made to return a `thrust::optional<block>`?

**Comment by kkraus14** [comment]
cc @shwina to review from the Python side

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 7 above)
Comment (codereport) [reply include/rmm/mr/device/pool_memory_resource.hpp:225]: And not that I want to recommend implementing are own `expected`-like type, but the `pair<error, result>` pattern is similar to:
- Rusts `Result` type: https://doc.rust-lang.org/std/result/
- Sy Brand's `tl::expected`: https://github.com/TartanLlama/expected

and I would say having a similar type would be much clearer than having raw `std::pair`s for this use case. 

This is obviously out of the scope of this PR - but just thought I would _throw_ it out there. :joy: And note `tl::expected` is based on the proposed [`std::expected`](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0323r3.pdf). And note note, `thrust::optional` is based on Sy Brand's `tl::optional`

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 8 above)
Comment (codereport) [reply include/rmm/mr/device/pool_memory_resource.hpp:184]: `thrust::optional` is better than exception throwing IMO, but I would prefer the `expected`/`Result` type: https://github.com/rapidsai/rmm/pull/601#discussion_r504104258

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 7 above)
Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp:225]: I don't remember saying anything about this. C++'s built-in exceptions sounds much better to me -- it is consistent with everything else in RMM.

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 8 above)
Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp:184]: No matter what, the upstream memory resource can throw an exception.

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 7 above)
Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp:225]: @codereport I also think you messed up the logic. :)  It returns remaining / 2 if size less than remaining, while the original returned remaining / 2 if size less than remaining / 2.

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 8 above)
Comment (jrhemstad) [reply include/rmm/mr/device/pool_memory_resource.hpp:184]: > No matter what, the upstream memory resource can throw an exception.

Yeah, the idea would be to catch the thrown exception inside `block_from_upstream` and return a `optional<block>` whose contents depend on the success of the allocation from the Upstream.

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 7 above)
Comment (codereport) [reply include/rmm/mr/device/pool_memory_resource.hpp:225]: > @codereport I also think you messed up the logic. :) It returns remaining / 2 if size less than remaining, while the original returned remaining / 2 if size less than remaining / 2.

I agree with your above statement - i think my thought is that they are equivalent. My visualization is the following:
```
r  = remaining
as = aligned size

< -------- r/2 ---------- r ---- >
            |             |
as =  r/2   |     as      |
```
If you look above, basically `remaining / 2` is a floor - which is what `std::max(remaining/2, aligned_size)`. My visualization might be wrong though. What do you think?

Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp:225]: Yeah, I think you are right. That is nicer.

File: include/rmm/mr/device/pool_memory_resource.hpp (hunk 2 above)
Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp]: I will try to add another test that sets no max and forces growth and if I can test this with an upstream that has a limited size I should be able to verify that growth succeeds and fails as expected.

**Comment by harrism** [comment]
I refactored all the growth and pool initialization stuff based on feedback. It's a bit more code, but it is DRYer.

**Comment by harrism** [comment]
@codereport @cwharris @jrhemstad I added a test that uses the new `limiting_resource_adaptor` to allow me to test the pool growth. This works because the pool gets no maximum size, but its upstream resource provides a limit, so we can use small sizes to test that pool growth works and fails as expected (without having to use real GPU memory sizes).

**Review by codereport — APPROVED** [review]
lgtm :+1:

**Comment by harrism** [comment]
@cwharris Up to you now.

**Review by cwharris — APPROVED** [review]
lgtm :rocket:

**Comment by cwharris** [comment]
Apparently I am not a python code owner for rmm! :)

**Review by shwina — APPROVED** [review]
LGTM!

**Comment by harrism** [comment]
Thanks @shwina
